### ISSUE TITLE
[CLI] Sweep launches multiple runs of the same parameter(s)

### ISSUE THREAD

**Issue by jmuchovej** [wandb/client#2611]
**Description**
Running multiple agents in parallel creates lots of repeated work, even hours after the configuration has been searched.

**Wandb features**
`wandb sweep` and `wandb agent`

**How to reproduce**
I’m not entirely sure how to guarantee a reproduction… 😕 I’ve been running the agents on a HPC.
1. The script’s I’m using are for ongoing research, which I’m willing to privately share, but not publicly. 😕 
2. All that I had to do was submit a sweep configuration, then run 20 agents. 

I don’t have a copy of the run folder at the moment, but I can see what I can do about getting copies of the multiple runs.

**Environment**
- OS: CentOS 7 (HPC environment)
- Environment: Folded below.

  <details>
    <summary> env.yml </summary>

    ```yaml
    name: project-name
    channels:
    - defaults
    - conda-forge
    - anaconda
    - pytorch
    dependencies:
    - black=20.8b1
    - bokeh=2.3.3
    - click=8.0.1
    - dask=2021.7.0
    - dask-jobqueue==0.7.3
    - distributed=2021.7.0
    - hydra-core=1.1.0
    - isort=5.9.2
    - jedi=0.18.0
    - knockknock=0.1.8.1
    - matplotlib=3.3.4
    - numpy=1.20.3
    - omegaconf=2.1.0
    - pandas=1.2.5
    - pip=21.1.3
    - pycodestyle=2.7.0
    - pydocstyle=5.1.1
    - pyflakes=2.3.1
    - pylama=7.7.1
    - pylint=2.9.3
    - python=3.9.5
    - pytorch-lightning=1.3.8
    - pytorch=1.8.1
    - scikit-learn=0.24.2
    - scipy=1.6.2
    - seaborn=0.11.1
    - tqdm=4.59.0
    - wandb=0.11.0
    variables:
      # MKL_DEBUG_CPU_TYPE: "5"  # Only necessary for AMD Ryzen CPUs
      # Choose one for determinism
      CUBLAS_WORKSPACE_CONFIG: ":4096:8"
      # CUBLAS_WORKSPACE_CONFIG: ":16:8"
  ```

  </details>
- Python Version: 3.9.5

Edit 1: Adding `env.yml`. Corrected the Python version.

**Comment by dannygoldstein** [comment]
Hi @jmuchovej, I understand that you can't share your training scripts, but could you share a example sweep config so we can attempt to reproduce?

**Comment by jmuchovej** [comment]
```yaml
program: main.py
description: "Description"
project: "project-name"
command:
- ${env}
- ${interpreter}
- ${program}
- mode=train
- +hydra=train
- +model=model
- +dataset=dataset
- experiment.name="some-name"
- ${args_no_hyphens}

# I should have specified grid here, but I was originally planning to have dozens
#   of parameters that, when crossed, would have taken the rest of the year to
#   compute. :/
method: bayes
metric:
  goal: minimize
  name: loss/total
parameters:
  seed:
    min: 21
    max: 84
    distribution: int_uniform
```

**Comment by jmuchovej** [comment]
Here's a screenshot of one of the worst "offenders" (9 repeated runs for `seed=57`).
![image](https://user-images.githubusercontent.com/5000729/131352875-0ad579ea-7050-4a31-a581-f4109d737046.png)

**Comment by jmuchovej** [comment]
I've managed to pull the logs from the HPC environment, so let me know if having those would be helpful. 🙂

**Comment by dannygoldstein** [comment]
Hi @jmuchovej,

I looked into this, and I don't think the repetition of parameter values is indicative of an underlying issue with the sweeps logic here. There are a few things to keep in mind:

1. You are using `method=bayes`, which to zeroth order, finds local optima in `loss/total` and, after running for some time, may  tend to concentrate suggestions around those optima. In a case like the one you show above of a sweep with a single int_uniform parameter with only ~60 potential values, it is not surprising that you will get certain values repeated that may correspond to local optima of the loss surface. If you are interested in exploring the parameter space instead of converging to an optimum, I would recommend using another sweep type for your search. The most efficient may be grid, since you have such a small number of options. But using `random` would also guarantee less concentration around a specific parameter value in the suggestions
2. I took a look at some of your sweep runs, and it looks like for a given value of seed there is some non-determinism in the values of `loss/total`. E.g., for the same value of seed = 23, there is a spread in values of `loss/total`, (from 4.26  to 5.15 at a quick first glance). So it looks like for a fixed seed there is a spread in loss values, the successive effect of which  may cause the optimizer to keep returning to values it has already seen.

**Comment by jmuchovej** [comment]
Hi @dannygoldstein, thanks for clarifying! This all makes more sense.

In hindsight, I should have realized this, but I think it _might_ make sense to add a note of this sort of behavior to the [docs][sweep-docs].

[sweep-docs]: https://docs.wandb.ai/guides/sweeps/configuration#method
